pkg/controller/controller_instance.go

```diff
@@ -281,6 +281,7 @@ func (c *controller) reconcileServiceInstance(instance *v1beta1.ServiceInstance)
 // more processing needed).
 func (c *controller) initObservedGeneration(instance *v1beta1.ServiceInstance) (bool, error) {
 	if instance.Status.ObservedGeneration == 0 && instance.Status.ReconciledGeneration != 0 {
+		instance = instance.DeepCopy()
```

Comment:
We do `DeepCopy` at the beginning of every `reconcileServiceInstance`/`reconcileServiceInstanceUpdate`/`reconcileServiceInstanceDelete` function (after checking that there's a work to do), but these two `init*` methods are invoked directly from the `reconcileServiceInstance` method.

Would it make sense to just make a deep copy in the beginning of `reconcileServiceInstance` method? I get that sometimes it will mean that we will make a deep copy while there is no status update, but I think that this overhead might be negligible.
The only case when this extra memory footprint can be significant is when we restart `controller-manager` pod (e.g. update Service Catalog Docker image), and reprocess all already processed resources.

Reply:
^I was trying to get to this point. I kind of prefer the simplicity. Unless we can create some copy-on-write impersonator for everything, but that's way out of scope.